### DISCUSSION TITLE
SpencerEverly Plans

### DISCUSSION THREAD

**Discussion by P-Star7** [Wohlstand/TheXTech#461] [Planning]
Here is the place where me and others can discuss the changes that SpencerEverly has proposed and whether or not we think they are good for TheXTech.

Original Post:
> This is basically a pull request with many new things surrounding my own edits. Nothing major, unless Wohlstand wants to add some of this stuff over officially to TheXTech, then I won't mind.
> 
> I only tested all this on Windows, so I'm not sure if compatibility for any other platforms exist or not.
> 
> Plans (Not specifying whether they're long term or not):
> 
> compat.ini: configuration-string = Whether it's using compat.ini or not
> 
> * [x]  Add player hammer throw sounds
> * [ ]  Add combo sounds (This needs more work tbh)
> * [ ]  Add every other sound from SMAS++ (An SMBX2 episode I'm working on)
>   
>   * [ ]  bowser-fireball (Bowser fireball for SMB1 Bowser and SMB3 Bowser)
>   * [ ]  fireball2 (Separate Fireball when a Venus Fire Trap spits out a fireball)
>   * [ ]  p-wing (For Leaf and Tanooki suits)
>   * [ ]  smb2-hitenemy (For when you hit projectiles on SMB2 enemies)
>   * [ ]  hp-get and hp-max sounds (HP indicators for Toad, Peach, and Link)
>   * [x]  0up (For the Infinity life mode, see below)
>   * [ ]  fireball-hit (When a fireball collides with a wall)
>   * [ ]  ice-melt (When a throw block melts)
> * [x]  Disable the disabling of saving while cheating (compat.ini: enable-anti-cheat)
> * [x]  Add an "exit to map/hub" option on the pause menu, and make it so you can save anywhere on the pause menu, even on a level (compat.ini: pause-and-exit-to-map, save-anywhere)
> * [x]  Make it so you can change characters on a level, not just on the map (compat.ini: switch-characters-on-the-fly)
> * [ ]  Add some X2 exclusive cheats
>   LEVEL SPECIFIC
>   
>   * [x]   getmeouttahere (Exits the level without winning)
>   * [x]  foundmycarkeys (Instantly grants an keyhole exit)
>   * [x]  mylifegoals (Instantly grants a SMW goal tape exit)
>   * [x]  mysteryball (Instantly grants a SMB3 goal orb exit)
>   * [x]  itsvegas (Instantly grants a SMB3 roulette exit)
>   * [ ]  getdemstars (Causes the player to teleport around the level, collecting all obtainable stars)
>   * [x]  newleaf (Deactivates all active cheats)
>   * [x]  holytrinity (Activates shadowstar, donthurtme, jumpman)
>   * [x]  passerby (Same as above)
>   * [x]  theessentials (Activates sonicstooslow, shadowstar, donthurtme, jumpman)
>   * [x]  theessenjls (Same as above)
>   * [ ]  waitinginthesky (Changes the starman music to [this](https://youtu.be/r-3kbvMSeLw?t=52), duration also gets affected, will be used when Starman from the plan list gets added)
>   * [ ]  thestarmen (Grants the starman effect, will be used when Starman from the plan list gets added)
>   * [ ]  3player (Changes to 3 player mode, will be used when 4 player multiplayer is fully implemented)
>   * [ ]  4player (Changes to 4 player mode, will be used when 4 player multiplayer is fully implemented)
> * [ ]  Add SMAS++/SEE-exclusive cheats (Certain ones)
>   LEVEL SPECIFIC
>   
>   * [ ]  iwannabootbackhome (Loads the Hub level if one exists. If none exists nothing will happen)
>   * [ ]  sherbertsmiddlenameistoto (Kills all players after typing)
>     WORLD SPECIFIC
>   * [ ]  parkmetoalevelunknown (Causes the player to teleport to a random level location, where it automatically unlocks the level after teleporting)
> * [ ]  Add a game over screen, with a menu to continue (Which will boot back to the world map refilling your lives), or exit to the main menu
> * [ ]  Add a life system, where you can choose between Normal (SMBX), Double (999 Lives Max), King (1110 lives, with three crowns), Triple (9999), or Infinity (No lives set, you can die as much as you want and still continue)
> * [ ]  Add an additional event option to stop an current running event (Will be useful for certain things)
> * [ ]  Add the Starman NPC and certain things for it (NPC 293, may be long term though)
> * [ ]  Add 4 player multiplayer
> * [ ]  Add the first 2 X2 characters onto TheXTech (Mega Man and Wario), which will be a test to see if I can analyze the code

**Comment by P-Star7** [comment]
> compat.ini: configuration-string = Whether it's using compat.ini or not
> 
> * [x]  Add player hammer throw sounds
> * [ ]  Add combo sounds (This needs more work tbh)
> * [ ]  Add every other sound from SMAS++ (An SMBX2 episode I'm working on)
>   
>   * [ ]  bowser-fireball (Bowser fireball for SMB1 Bowser and SMB3 Bowser)
>   * [ ]  fireball2 (Separate Fireball when a Venus Fire Trap spits out a fireball)
>   * [ ]  p-wing (For Leaf and Tanooki suits)
>   * [ ]  smb2-hitenemy (For when you hit projectiles on SMB2 enemies)
>   * [ ]  hp-get and hp-max sounds (HP indicators for Toad, Peach, and Link)
>   * [x]  0up (For the Infinity life mode, see below)
>   * [ ]  fireball-hit (When a fireball collides with a wall)
>   * [ ]  ice-melt (When a throw block melts)
All good ideas. Here are some additional suggestions:

- Also use the 0up sound when a player has collected a 1up/2up/3up/5up when they already have 99/999/1110/9999 lives.
- Special sound for the SMB2 Mushroom when Mario and Luigi collect it while in forms 3-7 (cloning their powerup into their reserve box)
- Dry Bones uncrumble sound (reverse of the crumble sound, is inspired by NSMB)
- A unique Note Block bounce sound (like in NSMBWii/NSMB2)
- Big Boo death sound (like when it falls off the screen in Donut Secret House)
- Max running speed (This might get annoying; DEFINITELY make this a separate config from the main "extra sounds" config)
- SMB3 Bowser Death (separate from SMB1 Bowser's death, because SMB3 should make the pit falling sound and SMB1 Bowser shouldn't. This sound also shouldn't be used when SMB3 Bowser dies in lava, only when he falls in a pit or is defeated by Link's sword or NPCs in such a way that his effect is made and he falls off the screen, because he also makes this sound in SMB3 if you kill him with fireballs or hammers.)
- SMB2 Heart (for when Peach and Toad are at 2 HP and collect a mushroom, so they now have 3 hearts but aren't powering up.)
- Link to the Past Heart Fill (for when Link picks up a heart when he has 1 or 2 HP)
- Link to the Past Throw (for when Link throws bombs)
- Link to the Past Bomb Place (for when Link places down a bomb while ducking)
-SM64 Star Grabbed - When a star is obtained (is not the music, is the sparkle sound when Mario grabs the star and he is falling)
- I think LTTP Fire Rod and Ice Rod sounds are already added by Wohlstand but check those just in case.
- The SMB2 Projectile Hit sound also plays when enemies are defeated in SMB2 by using a POW Block. It's up to you whether it should apply to only SMB2 enemies here or all enemies.
- Birdo is a unique case. She plays the SMB2 Projectile and Boss Hit sounds the first two hits, and the Boss Hit and Spitting Orb sounds on the third hit.
> * [x]  Disable the disabling of saving while cheating (compat.ini: enable-anti-cheat)

This is actually required for some compatibility. SMBX 1.1.x allowed cheating without consequences, so SMBX 1.1.x episodes should allow it too.
> * [x]  Add an "exit to map/hub" option on the pause menu, and make it so you can save anywhere on the pause menu, even on a level (compat.ini: pause-and-exit-to-map, save-anywhere)

The save-anywhere option is again required for SMBX 1.1.x compatibility. Though I don't really get its use since SMBX seems to save automatically whenever you get an exit or a star, which are the only things currently recorded in the save files. But this is a good thought in case it changes later.
> * [x]  Make it so you can change characters on a level, not just on the map (compat.ini: switch-characters-on-the-fly)

Good
> * [ ]  Add some X2 exclusive cheats
>   LEVEL SPECIFIC
>   
>   * [x]   getmeouttahere (Exits the level without winning)
>   * [x]  foundmycarkeys (Instantly grants an keyhole exit)
>   * [x]  mylifegoals (Instantly grants a SMW goal tape exit)
>   * [x]  mysteryball (Instantly grants a SMB3 goal orb exit)
>   * [x]  itsvegas (Instantly grants a SMB3 roulette exit)
>   * [ ]  getdemstars (Causes the player to teleport around the level, collecting all obtainable stars)
>   * [x]  newleaf (Deactivates all active cheats)
>   * [x]  holytrinity (Activates shadowstar, donthurtme, jumpman)
>   * [x]  passerby (Same as above)
>   * [x]  theessentials (Activates sonicstooslow, shadowstar, donthurtme, jumpman)
>   * [x]  theessenjls (Same as above)
>   * [ ]  waitinginthesky (Changes the starman music to [this](https://youtu.be/r-3kbvMSeLw?t=52), duration also gets affected, will be used when Starman from the plan list gets added)
>   * [ ]  thestarmen (Grants the starman effect, will be used when Starman from the plan list gets added)
>   * [ ]  3player (Changes to 3 player mode, will be used when 4 player multiplayer is fully implemented)
>   * [ ]  4player (Changes to 4 player mode, will be used when 4 player multiplayer is fully implemented)
> * [ ]  Add SMAS++/SEE-exclusive cheats (Certain ones)
>   LEVEL SPECIFIC
>   
>   * [ ]  iwannabootbackhome (Loads the Hub level if one exists. If none exists nothing will happen)
>   * [ ]  sherbertsmiddlenameistoto (Kills all players after typing)
>     WORLD SPECIFIC
>   * [ ]  parkmetoalevelunknown (Causes the player to teleport to a random level location, where it automatically unlocks the level after teleporting)

Good
> * [ ]  Add a game over screen, with a menu to continue (Which will boot back to the world map refilling your lives), or exit to the main menu

Possibly unnecessary
> * [ ]  Add a life system, where you can choose between Normal (SMBX), Double (999 Lives Max), King (1110 lives, with three crowns), Triple (9999), or Infinity (No lives set, you can die as much as you want and still continue)

Seems a little overkill except for Infinity Lives.
> * [ ]  Add an additional event option to stop an current running event (Will be useful for certain things)

See #101 and #100 for additional ideas here. This is a great idea.
> * [ ]  Add the Starman NPC and certain things for it (NPC 293, may be long term though)

This is a good idea. On the forums, there is a [thread](https://wohlsoft.ru/forum/viewtopic.php?f=94&t=4456) whose vote stated that NPCs do not have to expressly match SMBX2 NPCs 1:1, in behaviour or in order. I think the devs have stated that it is planned that any npc with a .lua file (such as in an episode, level, or when playing either the SMBX2 basegame assets) will have its vanilla NPC's behaviour completely overwritten, so it shouldn't be a worry for the basegame NPCs to be different.
> * [ ]  Add 4 player multiplayer

A noble idea. This will be very hard to make compatible with multi-res though, which is highly recommended so that there is more screen space for the four players.
> * [ ]  Add the first 2 X2 characters onto TheXTech (Mega Man and Wario), which will be a test to see if I can analyze the code

This might not be such a good idea. I think a better and more useful method of analyzing the code would be to make more player parameters editable in the Player Calibrator for the default five players. We can discuss more what this could possibly entail later if you think it's a good idea. player-#.inis for the map would also be a good idea, as well as some sort of "mario-0" or putting this information in player-#.ini about information that affects all seven forms, such as the displayed name in the Save File/Battle Mode menus.
It might also be a good idea to firstly implement a "character clones" system similar to SMBX2, where all five of the player slots can be set to be based off of the Mario, Luigi, Peach, Toad, or Link AIs, so that for example there can be episodes or battle mode levels with four players that all play exactly like Mario but look different to make it more fair. This also would be a good framework for when Lua is implemented so that characters can be used as a base similar to SMBX2. Setting this groundwork now would make adding these truly new characters easier in the long run, and also Lua-compatible.

**Comment by ds-sloth** [comment]
I agree that most of these seem like good feature ideas.

Some minor notes: I don't see any reason to add additional SFX, and I think individual SFX should be reviewed on a case-by-case basis. Our engine officially targets the SMBX64 experience, with some inspiration from X2 and 38A. We might draw inspiration from how other games implement features we're already interested in, but we're not turning to them to look for features to implement. Making an SMAS++-like episode is a fun idea, but features specifically for something like SMAS++ do not belong in the core game engine. When we add Lua scripting, we will certainly make it possible to load and play new audio effects.

> Make it so you can change characters on a level, not just on the map (compat.ini: switch-characters-on-the-fly)

This was discussed and fully implemented when we developed the Abstract Controls update over a year ago, but we intentionally decided to disable it in most cases. @0lhi may remember some of the reasoning behind this?

================

Generally, please remember that the core developer team has some specific priorities for the game engine overall (unicode fonts (just got added to the main branch today), multires, localizations, config/compat, scripting), and time spent on various smaller features detracts from our time on these bigger issues. Additionally, high performance and engine simplicity are both primary aims of TheXTech (more so than for other SMBX-like engines). Thus, every feature needs to meet a high bar of user benefit relative to its implementation or review cost, memory and processing cost, and game complexity cost.

I would gently request that people interested in newly contributing code to start with smaller modifications than 4P multiplayer. Given that @Wohlstand and I both have expertise with the relevant parts of the codebase, the difficulty with 4P would not the implementation (we laid a lot of the technical groundwork during the Abstract Controls update in 1.3.6), but in designing a good user experience. In our Discord and GitHub over the last year, we've had many discussions about various aspects of how >2P multiplayer should work, but we still don't have a solid plan. If you're really excited to get >2P multiplayer going, please contribute to #46 with design ideas and we can talk more about how to implement this exciting feature.

I'll be happy to review PRs for any of the features mentioned here, except for new characters. That can stay in your fork if you're really excited to make it, but I wouldn't want to import it.

**Reply by Wohlstand** [reply]
Adding new SFXes (like a combo hit sounds) is absolutely fine, there is a note that fallback rules must be added to ensure that older asset packs will behave as-is. I already added a pile of new sounds by myself. But again, add them with a fallback rules.

**Reply by ds-sloth** [reply]
I don't agree with this in general. Loading the SFX already takes 95% of the load time on Wii if the user has preconverted their graphics assets.

I'm definitely happy to add generally useful SFX, like combos. (That said, they should be implemented by playing the existing SFX at a higher pitch.) But I don't have any interest in implementing a game-wide new SFX for character 4's double jump. That would increase the game size and make life more difficult for asset pack maintainers with very little meaningful benefit to users.

In general, the development team should discuss the benefits and costs of each new feature instead of adding them by default, especially when we already have a backlog of important features that the whole team agrees are important.

**Reply by LoveBodhi** [reply]
I don't even think so. Since new SFX's are very useful for game engine improvement, but consider about device limitations, a "Mode Switch" system should be planned.

**Reply by ds-sloth** [reply]
It's not just a device limitation, it's also the limitation of our developer time. I'd rather focus on impactful features like enabling scriptable SFX via Lua (so we *shouldn't* have to add SFX to the engine every time someone has an idea for something they'd like a separate SFX for), rather than adding many SFX and continually updating the sound libraries and assets.

**Reply by LoveBodhi** [reply]
So, for example, how about separate Thwomp falling and POW block SFX? (This is in my [Modern Sound Pack](https://wohlsoft.ru/forum/viewtopic.php?f=29&t=4239), which features many type of Sound Effects)

**Reply by Wohlstand** [reply]
Also, speaking about SFX loading/playing... Currently, they get fully pre-decoded to be loaded, which takes a lot of time on weak devices. And now, I want to try load them as music and play them via multi-music API. The disadvantage of this method: is the impossibility of playing the same sound multiple times in parallel without loading more duplicated instances of the same file. So, I'll try to make next:
- On the rest of the devices, prefer the old method as this guarantees the low latency and ability to play the same SFX multiple times
- On weak devices, prefer the multi-music method to load sounds as `Mix_Music*` instances, and play them via MultiMusic API.

**Reply by 0lhi** [reply]
> This was discussed and fully implemented when we developed the Abstract Controls update over a year ago, but we intentionally decided to disable it in most cases. @0lhi may remember some of the reasoning behind this?

Been a while, but I think this was since we decided to not have a life punishment for using Drop/Add inside a level because some parts are impossible to beat in Multiplayer. To compensate, character changing was restricted to Hub/Map.

**Comment by LoveBodhi** [comment]
I forgot to say, I should suggest a "Customizable Sound Role" system.
